File: Data Structures & Algorithms/search-for-word-ii/submission-0.java
class Trie {
    String word;
    boolean eow;
    Trie[] ch;

    Trie() {
        word = "";
        eow = false;
        ch = new Trie[26];
    }

    public void addWord(String w) {
        Trie curr = this;

        for (char c : w.toCharArray()) {
            int i = c - 'a';
            if (curr.ch[i] == null) {
                curr.ch[i] = new Trie();
            }
            curr = curr.ch[i];
        }

        curr.eow = true;
        curr.word = w;
    }
}

class Solution {
    Set<String> res;
    boolean[][] visit;

    public List<String> findWords(char[][] board, String[] words) {
        Trie root = new Trie();

        for (String w : words) {
            root.addWord(w);
        }

        int ROWS = board.length;
        int COLS = board[0].length;

        res = new HashSet<>();
        visit = new boolean[ROWS][COLS];

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                dfs(board, r, c, root);
            }
        }

        return new ArrayList<>(res);
    }

    private void dfs(char[][] board, int r, int c, Trie node) {
        if (r < 0 || c < 0 || r >= board.length || c >= board[0].length)
            return;

        if (visit[r][c])
            return;

        char ch = board[r][c];
        Trie next = node.ch[ch - 'a'];

        if (next == null)
            return;

        if (next.eow) {
            res.add(next.word);
        }

        visit[r][c] = true;

        dfs(board, r + 1, c, next);
        dfs(board, r - 1, c, next);
        dfs(board, r, c + 1, next);
        dfs(board, r, c - 1, next);

        visit[r][c] = false;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-0.cpp
class Solution {
public:
    int ROWS, COLS;

    bool dfs(vector<vector<char>>& brd, string& w, int r, int c, int i) {
        if (i == w.size()) {
            return true;
        }

        if (r < 0 || r >= ROWS || c < 0 || c >= COLS ||
            brd[r][c] != w[i] || brd[r][c] == '#') {
            return false;
        }

        char temp = brd[r][c];
        brd[r][c] = '#';

        bool res =
            dfs(brd, w, r + 1, c, i + 1) ||
            dfs(brd, w, r - 1, c, i + 1) ||
            dfs(brd, w, r, c + 1, i + 1) ||
            dfs(brd, w, r, c - 1, i + 1);

        brd[r][c] = temp;

        return res;
    }

    bool exist(vector<vector<char>>& board, string word) {
        ROWS = board.size();
        COLS = board[0].size();

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (dfs(board, word, r, c, 0)) {
                    return true;
                }
            }
        }

        return false;
    }
};